### PR TITLE
Add no-op controlplane controller to the local provider

### LINKED ISSUE #21: Extensibility for custom Shoot control plane components
[GEP-1](https://github.com/gardener/gardener/blob/master/docs/proposals/01-extensibility.md) proposes a mechanism for building a custom Shoot control plane based on `Generic` resources. This has the drawback that an extension controller handling custom control plane is receiving a lot of information that may not be relevant (e.g. the entire shoot specification), while at the same time not receiving information that may be relevant or even needed (e.g. the `InfrastructureStatus`).

This issue proposes a somewhat more elaborate and specific mechanism for adding custom Shoot control plane components. Unlike the ideas outlined in [Control plane extension](https://github.com/gardener/gardener-extensions/issues/20), this proposal builds upon the principles and mechanisms already specified in GEP-1, elaborating on them only to the point to successfully meet the needs of the currently known cloud providers (including an internal SAP provider known as SEN). Instead of focusing on flexibility and composability for future use cases, it aims to be as simple and compatible with the ongoing work on GEP-1 as possible.

## Adding Custom Shoot Control Plane Components

To support adding custom control plane components, the following Shoot definition contains a new `controlPlane` section:

```yaml
apiVersion: gardener.cloud/v1alpha1
kind: Shoot
metadata:
  name: sen-1
  namespace: garden-dev
spec:
  cloudProfileName: sen
  secretBindingName: core-sen
  cloud:
    type: sen
    region: sen
    ...
    controlPlane:
      providerConfig:
        apiVersion: sen.extensions.gardener.cloud/v1alpha1
        kind: ControlPlaneConfig
        # SEN-specific config ...
```

As a result, at a certain point in the Gardener flow right after the Kubernetes API server is ready, Gardener writes a custom resource of type `ControlPlane` to the shoot namespace. The extension controller acts upon this resource by deploying all custom Shoot control plane components, and eventually updates its status:

```yaml
apiVersion: extensions.gardener.cloud/v1alpha1
kind: ControlPlane
metadata:
  name: control-plane
spec:
  type: sen
  shoot: sen-1
  region: sen
  providerConfig:
    apiVersion: sen.extensions.gardener.cloud/v1alpha1
    kind: ControlPlaneConfig
    # SEN-specific config (specified in the shoot) ...
  infrastructureProviderStatus:
    apiVersion: sen.extensions.gardener.cloud/v1alpha1
    kind: InfrastructureStatus
    # SEN-specific config (returned by SEN Infrastructure controller) ...
  secretRef:
    name: cloudprovider
    namespace: shoot--dev--sen-1
  apiServerAddress: 1.2.3.4
status:
  observedGeneration: ...
  state: ...
  lastError: ..
  lastOperation: ...
  providerStatus:
    apiVersion: sen.extensions.gardener.cloud/v1alpha1
    kind: ControlPlaneStatus
    # SEN-specific status
```

As it is, this proposal does not address the "exposure" use case, i.e. how the control plane (`kube-apiserver`) of the shoot is exposed to end-users (today this is done via a load balancer service). This depends on the Seed's cloud provider. For some cloud providers (e.g. AWS), additional control plane components are needed as well. We should probably introduce a different resource, `ControlPlaneExposure`, for this case. This will be proposed in a different issue.

### PR DESCRIPTION
**Pull request by stoyanr**
**What this PR does / why we need it**:
Adds a no-op extension controller that handles ControlPlane resources to the local provider.

**Which issue(s) this PR fixes**:
Fixes #21.

**Special notes for your reviewer**:
This should now be mergeable.

**Release note**:

```improvement operator
Add no-op controlplane controller to the local provider
```

### CODE REVIEW

File: pkg/controller/controlplane/reconciler.go [hunk 1]
```diff
@@ -0,0 +1,181 @@
+package controlplane
```
Comment (rfranzke) [inline pkg/controller/controlplane/reconciler.go]: license header missing

Comment (stoyanr) [reply pkg/controller/controlplane/reconciler.go]: Fixed now

File: pkg/controller/controlplane/reconciler.go [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+	r.logger.Info("Starting the reconciliation of controlplane", "controlplane", cp.Name)
+	r.recorder.Event(cp, corev1.EventTypeNormal, EventControlPlaneReconciliation, "Reconciling the controlplane")
+	if err := r.actuator.Reconcile(ctx, cp, cluster); err != nil {
+		msg := "Error reconciling controlplane"
+		r.updateStatusError(ctx, extensionscontroller.ReconcileErrCauseOrErr(err), cp, operationType, msg)
+		r.logger.Error(err, msg, "controlplane", cp.Name)
+		return extensionscontroller.ReconcileErr(err)
+	}
+
+	if kutil.HasMetaDataAnnotation(&cp.ObjectMeta, gardencorev1alpha1.GardenerOperation, gardencorev1alpha1.GardenerOperationReconcile) {
```
Comment (rfranzke) [inline pkg/controller/controlplane/reconciler.go]: Can you remove this? It's not needed for the ControlPlane controller.

Comment (stoyanr) [reply pkg/controller/controlplane/reconciler.go]: Sure, done

**Pull request merged**